presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -0,0 +1,724 @@
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a > CAST(NULL AS DOUBLE)",
+                output(
+                        filter("CAST(NULL AS BOOLEAN)",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < CAST(NULL AS DOUBLE)",
+                output(
+                        filter("CAST(NULL AS BOOLEAN)",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a >= CAST(NULL AS DOUBLE)",
+                output(
+                        filter("CAST(NULL AS BOOLEAN)",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <= CAST(NULL AS DOUBLE)",
+                output(
+                        filter("CAST(NULL AS BOOLEAN)",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a IS DISTINCT FROM CAST(NULL AS DOUBLE)",
+                output(
+                        values("A")));
+    }
+
+    @Test
+    public void smokeTests()
+    {
+        // smoke tests for various type combinations
+        for (String type : Arrays.asList("SMALLINT", "INTEGER", "BIGINT", "REAL", "DOUBLE")) {
+            assertPlan(
+                    format("SELECT * FROM (VALUES TINYINT '1') t(a) WHERE a = %s '1'", type),
+                    anyTree(
+                            filter("A = TINYINT '1'",
+                                    values("A"))));
+        }
+
+        for (String type : Arrays.asList("INTEGER", "BIGINT", "REAL", "DOUBLE")) {
+            assertPlan(
+                    format("SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = %s '1'", type),
+                    anyTree(
+                            filter("A = SMALLINT '1'",
+                                    values("A"))));
+        }
+
+        for (String type : Arrays.asList("BIGINT", "DOUBLE")) {
+            assertPlan(
+                    format("SELECT * FROM (VALUES INTEGER '1') t(a) WHERE a = %s '1'", type),
+                    anyTree(
+                            filter("A = 1",
+                                    values("A"))));
+        }
+
+        assertPlan("SELECT * FROM (VALUES REAL '1') t(a) WHERE a = DOUBLE '1'",
+                anyTree(
+                        filter("A = REAL '1.0'",
+                                values("A"))));
+    }
+
+    @Test
+    public void testTermOrder()
+    {
+        // ensure the optimization works when the terms of the comparison are reversed
+        // vs the canonical <expr> <op> <literal> form
+        assertPlan("SELECT * FROM (VALUES REAL '1') t(a) WHERE DOUBLE '1' = a",
+                anyTree(
+                        filter("A = REAL '1.0'",
+                                values("A"))));
+    }
+
+    @Test
+    public void testNoEffect()
```

Comment:
Add a test case with `decimal(15,0)`, `double` and `nan` / `infinity` literals. Conversion from target type to source type will fail and the rule will bail out.

Reply:
`decimal(15, 0)` actually works fine. There's already a test for `decimal(16, 0)`

Reply:
>  Conversion from target type to source type will fail and the rule will bail out.

Unfortunately, not. I'd need a type that has no range defined for that to happen.

Reply:
To be precise, I mean how to test-cover https://github.com/prestosql/presto/pull/731/files#diff-c155c41e6fc860986d8a59bb77548dafR288

`decimal(15,0)` fits in double without loss (so `isInjectiveConversion` will not stop the code from proceeding)
Both decimal and double don't have ranges defined (Decimal could have one). So we should attempt literal-to-source-type cast.